From: Jasper Lievisse Adriaanse <j@jasper.la>
Date: Fri, 26 Feb 2021 15:21:20 +0100
Subject: Fix potential memory corruption with negative memmove() size
Origin: https://github.com/lz4/lz4/commit/8301a21773ef61656225e264f4f06ae14462bca7
Bug: https://github.com/lz4/lz4/pull/972
Bug-Debian: https://bugs.debian.org/987856
Bug-Debian-Security: https://security-tracker.debian.org/tracker/CVE-2021-3520
Bug-SUSE: https://bugzilla.suse.com/show_bug.cgi?id=1185438

[Salvatore Bonaccorso: Backport to 1.8.3:
 - context changes for various refactorings in LZ4_decompress_generic()
 - Add addtional NULL check following upstream changes in 8bea19d57c0d ("fixed
   minor cppcheck warnings in lib")
]
---
 lib/lz4.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/lib/lz4.c
+++ b/lib/lz4.c
@@ -1414,6 +1414,8 @@ LZ4_decompress_generic(
                  const size_t dictSize         /* note : = 0 if noDict */
                  )
 {
+    if ((src == NULL) || (outputSize < 0)) { return -1; }
+
     const BYTE* ip = (const BYTE*) src;
     const BYTE* const iend = ip + srcSize;
 
@@ -1421,7 +1423,7 @@ LZ4_decompress_generic(
     BYTE* const oend = op + outputSize;
     BYTE* cpy;
 
-    const BYTE* const dictEnd = (const BYTE*)dictStart + dictSize;
+    const BYTE* const dictEnd = (dictStart == NULL) ? NULL : dictStart + dictSize;
     const unsigned inc32table[8] = {0, 1, 2,  1,  0,  4, 4, 4};
     const int      dec64table[8] = {0, 0, 0, -1, -4,  1, 2, 3};
 
